crates/bevy_math/src/cubic_splines.rs

```diff
@@ -269,30 +294,349 @@ impl<P: Point> CubicBSpline<P> {
+    knots: Vec<f32>,
+}
+impl<P: Point> CubicNurbs<P> {
+    /// Build a Non-Uniform Rational B-Spline.
+    ///
+    /// If provided, weights must be the same length as the control points. Defaults to equal weights.
+    ///
+    /// If provided, the number of knots must be n + 4 elements, where n is the amount of control
+    /// points. Defaults to open uniform knots: [`Self::open_uniform_knots`]. Knots cannot
+    /// all be equal.
+    ///
+    /// At least 4 points must be provided, otherwise an error will be returned.
+    pub fn new(
+        control_points: impl Into<Vec<P>>,
+        weights: Option<impl Into<Vec<f32>>>,
+        knots: Option<impl Into<Vec<f32>>>,
+    ) -> Result<Self, CubicNurbsError> {
+        let mut control_points: Vec<P> = control_points.into();
+        let control_points_len = control_points.len();
+
+        if control_points_len < 4 {
+            return Err(CubicNurbsError::NotEnoughControlPoints {
+                provided: control_points_len,
+            });
+        }
+
+        let weights = weights
+            .map(Into::into)
+            .unwrap_or_else(|| vec![1.0; control_points_len]);
+
+        let mut knots: Vec<f32> = knots.map(Into::into).unwrap_or_else(|| {
+            Self::open_uniform_knots(control_points_len)
+                .expect("The amount of control points was checked")
+        });
+
+        let expected_knots_len = Self::knots_len(control_points_len);
+
+        // Check the number of knots is correct
+        if knots.len() != expected_knots_len {
+            return Err(CubicNurbsError::KnotsNumberMismatch {
+                expected: expected_knots_len,
+                provided: knots.len(),
+            });
+        }
+
+        // Ensure the knots are non-descending (previous elements is less than or equal
+        // to the next)
+        if knots.windows(2).any(|win| win[0] > win[1]) {
+            return Err(CubicNurbsError::DescendingKnots);
+        }
+
+        // Ensure the knots are non-constant
+        if knots.windows(2).all(|win| win[0] == win[1]) {
+            return Err(CubicNurbsError::ConstantKnots);
+        }
+
+        // Check that the number of weights equals the number of control points
+        if weights.len() != control_points_len {
+            return Err(CubicNurbsError::WeightsNumberMismatch {
+                expected: control_points_len,
+                provided: weights.len(),
+            });
+        }
+
+        // To align the evaluation behavior of nurbs with the other splines,
+        // make the intervals between knots form an exact cover of [0, N], where N is
+        // the number of segments of the final curve.
+        let curve_length = (control_points.len() - 3) as f32;
+        let min = *knots.first().unwrap();
+        let max = *knots.last().unwrap();
+        knots = knots
+            .into_iter()
+            .map(|k| k - min)
+            .map(|k| k * curve_length / max)
+            .collect();
```

Comment:
Something worries me that the scaling of the sample parameter (and perhaps generally non-uniformity) will affect the derivatives by some factor.
Although this happens before the matrices and coefficients are computed, so I think it does get properly reflected.

Probably beyond the scope of this PR, but I believe we will need a test that checks that the derivative functions (velocity and acceleration) match up with the numerical derivatives of position.

Reply:
If you look at the choi paper, it specifies the derivatives in terms of `u`, which is defined as `u=(t-t_i)/(t_i+1 - t_i)` in Eq. 6. This is equivalent to the normalized term here.

I agree on the test.